src/tasks/lm_parsing.py

```diff
@@ -0,0 +1,150 @@
+"""
+Task definitions for language modeling parsing tasks. Used for unsupervised parsing encoders.
+Long term dependencies for language modeling: sentences concatenated together seperated by <EOS> token.
+"""
+import json
+import logging as log
+import math
+import os
+
+from allennlp.training.metrics import Average
+from allennlp.data.token_indexers import SingleIdTokenIndexer
+
+# Fields for instance processing
+from allennlp.data import Instance, Token
+
+from ..utils.utils import process_sentence, truncate, load_tsv
+
+from typing import Iterable, Sequence, List, Dict, Any, Type
+
+from .tasks import SequenceGenerationTask
+from .tasks import sentence_to_text_field, atomic_tokenize
+from .tasks import UNK_TOK_ALLENNLP, UNK_TOK_ATOMIC
+from .lm import LanguageModelingTask
+from .registry import register_task
+
+
+class LanguageModelingParsingTask(LanguageModelingTask):
+    def count_examples(self):
+        """Computes number of samples.
+        Every example is made up of sentences concatenated together, capped by max_seq_len.
+        """
+        example_counts = {}
+        for split, split_path in self.files_by_split.items():
+            arr = [line.strip().split()+["<EOS>"] for line in open(split_path)]
+            allf = 0
+            for x in arr:
+                allf += len(x)
+            example_counts[split] = int(math.ceil(allf / self.max_seq_len))
+        self.example_counts = example_counts
+
+    def process_split(self, split, indexers) -> Iterable[Type[Instance]]:
+        """Process a language modeling split by indexing and creating fields.
+        Args:
+            split: (list) a single list of sentences
+            indexers: (Indexer object) indexer to index input words
+        """
+        def _make_instance(sent):
+            ''' Forward targs adds <s> as a target for input </s>
+            and bwd targs adds </s> as a target for input <s>
+            to avoid issues with needing to strip extra tokens
+            in the input for each direction '''
+            d = {}
+            d["input"] = sentence_to_text_field(sent[:-1], indexers)
+            d["targs"] = sentence_to_text_field(sent[1:], self.target_indexer)
+            d["targs_b"] = sentence_to_text_field([sent[-1]] + sent[:-2], self.target_indexer)
+            return Instance(d)
+        for sent in split:
+            yield _make_instance(sent)
+
+
+@register_task('wsj', rel_path='Penn/')
```

Comment:
WSJ or PTB would be a more informative dir name. It's never really called Penn, and Penn (the school) does a ton of stuff in NLP.

Reply:
Changed to WSJ.